### PR TITLE
[WIP] HPC improvements

### PR DESCRIPTION
**Pull request by ikervagyok**
###### Motivation for this change
Want to convince my colleagues, that `NixOS`/`NixOps` is OK for a test cluster setup which consists of a few AArch64 nodes. `LAMMPS` is the package I chose to give to my test users, that's why I've converted it to `CMake` and added a few extra `LAMMPS packages`

###### Things done
Updated stuff and added `libfabric`, `pmix` and `ucx`. Tested with 2 nodes, will go up to 4 tomorrow and maybe some x86 nodes in the next few days. 

@markuskowa since you seem to be at the center of NixOS-HPC what do you think about these changes? What should be made optional/reworked? I know It's not a polished PR, but I think can be made into something usable.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by costrouc** [comment]
Very happy to see that lammps finally has a standard build procedure with cmake!

File: nixos/modules/services/computing/slurm/slurm.nix [hunk 1]
```diff
@@ -355,6 +355,7 @@ in
         ExecStart = "${wrappedSlurm}/bin/slurmd";
         PIDFile = "/run/slurmd.pid";
         ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        LimitMEMLOCK = "infinity";
```
Comment (markuskowa) [inline nixos/modules/services/computing/slurm/slurm.nix]: Why is this necessary?

File: pkgs/development/libraries/openmpi/default.nix [hunk 2]
```diff
@@ -40,21 +52,28 @@ in stdenv.mkDerivation rec {
...
+    "--with-ofi=${libfabric}"
+    "--with-pmi=${pmix}"
+    "--with-pmix=${pmix} --enable-install-libpmix"
+    "--with-verbs=${rdma-core}"
+    "--with-ucx=${ucx}"
+    ]
+    ++ lib.optional (!cudaSupport) "--disable-mca-dso"
     ++ lib.optional isLinux  "--with-libnl=${libnl.dev}"
     ++ lib.optional enableSGE "--with-sge"
+    ++ lib.optional enableSLURM "--with-slurm"
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: This is the already the default.

File: pkgs/development/libraries/openmpi/default.nix [hunk 3]
```diff
@@ -40,21 +52,28 @@ in stdenv.mkDerivation rec {
...
   nativeBuildInputs = [ perl ];
 
-  configureFlags = with stdenv; lib.optional (!cudaSupport) "--disable-mca-dso"
+  configureFlags = with stdenv; [
+    "--with-cma"
+    "--with-hwloc=${hwloc.dev}"
+    "--with-ofi=${libfabric}"
+    "--with-pmi=${pmix}"
+    "--with-pmix=${pmix} --enable-install-libpmix"
+    "--with-verbs=${rdma-core}"
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: openmpi is already build against rdma-core. This line is not neccesary.

File: pkgs/development/libraries/openmpi/default.nix [hunk 4]
```diff
@@ -40,21 +52,28 @@ in stdenv.mkDerivation rec {
...
     ++ [ libevent hwloc ]
     ++ lib.optional (isLinux || isFreeBSD) rdma-core;
 
   nativeBuildInputs = [ perl ];
 
-  configureFlags = with stdenv; lib.optional (!cudaSupport) "--disable-mca-dso"
+  configureFlags = with stdenv; [
+    "--with-cma"
+    "--with-hwloc=${hwloc.dev}"
+    "--with-ofi=${libfabric}"
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: What benefit does libfabric bring (I am not familiar with libfabric)?

File: pkgs/development/libraries/openmpi/default.nix [hunk 5]
```diff
@@ -40,21 +52,28 @@ in stdenv.mkDerivation rec {
...
+    "--with-pmi=${pmix}"
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: I have tried to build with pmix in the past but could not get it work properly with slurm. Did you try if you can run openmpi binaries with `srun` instead of `mpirun`?
In general I would say pmix is not strictly necessary as openmpi already works nicely with slurm and detects the resource allocations properly. From how i understand pmix it is designed for large mpi jobs (>100 nodes).

File: pkgs/development/libraries/openmpi/default.nix [hunk 6]
```diff
@@ -40,21 +52,28 @@ in stdenv.mkDerivation rec {
...
+    "--with-pmix=${pmix} --enable-install-libpmix"
+    "--with-verbs=${rdma-core}"
+    "--with-ucx=${ucx}"
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: We (@mekeetsa) are testing UCX with infiniband at the moment (see also https://github.com/markuskowa/NixOS-QChem/blob/ucx/ucx/default.nix). Did you test it with an Infiniband configuration?

File: pkgs/development/libraries/openmpi/default.nix [hunk 7]
```diff
@@ -29,6 +33,14 @@ in stdenv.mkDerivation rec {
     sha256 = "0ms0zvyxyy3pnx9qwib6zaljyp2b3ixny64xvq3czv3jpr8zf2wh";
   };
 
+  # TODO: Remove these, when a new release gets out
+  # https://github.com/open-mpi/ompi/commit/526775dfd7ad75c308532784de4fb3ffed25458f.patch
+  # https://github.com/open-mpi/ompi/commit/a3026c016a6a8be379f62585b6ddc070175c8106.patch
+  patches = [
+    ./ucx-1.8.patch
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: There is only UCX 1.7?

File: pkgs/development/libraries/openmpi/default.nix [hunk 8]
```diff
@@ -29,6 +33,14 @@ in stdenv.mkDerivation rec {
...
+    ./ucx-1.7.patch
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: According to [upstream UCX](https://github.com/openucx/ucx/wiki/OpenMPI-and-OpenSHMEM-installation-with-UCX#running-open-mpi-with-ucx), configuring openmpi with `--enable-mca-no-build=btl-uct` solves the problem with no impact on performance.

File: pkgs/servers/computing/slurm/default.nix [hunk 9]
```diff
@@ -2,10 +2,20 @@
...
 , ncurses, libmysqlclient, gtk2, lua, hwloc, numactl
 , readline, freeipmi, libssh2, xorg, lz4
+, buildEnv, libibmad, libibumad, pmix, ucx
 # enable internal X11 support via libssh2
 , enableX11 ? true
 }:
 
+let
+
+  ofed = buildEnv {
```
Comment (markuskowa) [inline pkgs/servers/computing/slurm/default.nix]: `libibmad` and `libibumad` should be included in `rdma-core` now. Those packages are outdated.

**Review by markuskowa — CHANGES_REQUESTED** [review]
Hi @ikervagyok. Thank you for the contribution. It's nice to see that there is more and more HPC activity with Nix. As a general comment: this is a big PR that addresses several different issues.
It may be better to split it it up in several PRs (e.g. lammps, openmpi, slurm). It would also simplify the review if you could briefly comment how you have tested the additional features. 
In my experience it is one thing to get things to compile but a completely separate thing to make them work. There is usually some fine tuning involved. I test drive new packages in my own [overlay](https://github.com/markuskowa/NixOS-QChem) before I upstream them (feel free to copy from the overlay what you need).

File: pkgs/servers/computing/slurm/default.nix (hunk 9 above)
Comment (ikervagyok) [reply pkgs/servers/computing/slurm/default.nix]: thought that too, but it didn't work for some reason. will test it again.

File: pkgs/development/libraries/openmpi/default.nix (hunk 7 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: yes, but the patch is in preparation for the 1.8 release and they only later realized that 1.7 was broken as well.

File: pkgs/development/libraries/openmpi/default.nix (hunk 8 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: i usually prefer to build the package as a whole, to give all options to users - but i've tested `--enable-mca-no-build=btl-uct` and it also works that way.

File: pkgs/development/libraries/openmpi/default.nix (hunk 6 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: yes, i've tested it and it seems to work (setting up the env with `OMPI_MCA_btl=^openib,uct,tcp` `UCX_TLS=ib,sm,self`)

File: pkgs/development/libraries/openmpi/default.nix (hunk 5 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: you're right, i've thought it worked, but i probably mistakenly tested mpirun multiple times?! will look into this.

File: pkgs/development/libraries/openmpi/default.nix (hunk 4 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: I think OmniPath needs `libfabric` - can't try though, since I have no OmniPath machine in a test setup

File: pkgs/development/libraries/openmpi/default.nix (hunk 3 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: yeah, i've got a bit verbose in some places. will remove that line.

File: pkgs/development/libraries/openmpi/default.nix (hunk 2 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: will revert the `enableSLURM` option

File: nixos/modules/services/computing/slurm/slurm.nix (hunk 1 above)
Comment (ikervagyok) [reply nixos/modules/services/computing/slurm/slurm.nix]: This fixed the issue of slurm not being able to use mpirun, while it worked in a shell. I'll try again to see if i've messed up my testing.

File: pkgs/servers/computing/slurm/default.nix (hunk 9 above)
Comment (markuskowa) [reply pkgs/servers/computing/slurm/default.nix]: It would be good if that works. Eventually `libibmad` and `libibumad` should be removed from nixpkgs.

File: pkgs/development/libraries/openmpi/default.nix (hunk 7 above)
Comment (markuskowa) [reply pkgs/development/libraries/openmpi/default.nix]: There should not be unnecessary patches. If it becomes relevant with the next release then it should be added. Consider also to use `fetchPatch` for patches that can be fetched online.

File: pkgs/development/libraries/openmpi/default.nix (hunk 8 above)
Comment (markuskowa) [reply pkgs/development/libraries/openmpi/default.nix]: The solution to disable the uct btl seems to be more simple thus I would prefer that one (unless there is a motivation to activate it).

File: pkgs/development/libraries/openmpi/default.nix (hunk 4 above)
Comment (markuskowa) [reply pkgs/development/libraries/openmpi/default.nix]: I would prefer to remove it then.

File: pkgs/development/libraries/openmpi/default.nix (hunk 8 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: should be resolved with the `fetchpatch` approach from above and can be removed when the next openmpi release lands.

File: pkgs/development/libraries/openmpi/default.nix (hunk 4 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: as a dependency, or the whole package?

File: pkgs/development/libraries/openmpi/default.nix [hunk 10]
```diff
@@ -29,6 +30,12 @@ in stdenv.mkDerivation rec {
     sha256 = "0ms0zvyxyy3pnx9qwib6zaljyp2b3ixny64xvq3czv3jpr8zf2wh";
   };
 
+  # TODO: Remove these, when a new release gets out
+  patches = [
+    (fetchpatch { name = "ucx-1.8-compat-1"; url = "https://github.com/open-mpi/ompi/commit/526775dfd7ad75c308532784de4fb3ffed25458f.patch"; sha256 = "1zfr7ym4303dsnnjqfc1qbnfs0izd8bazxfl57p3dhqsiznd734r"; })
```
Comment (markuskowa) [inline pkgs/development/libraries/openmpi/default.nix]: I am still confused why there are patches for UCX 1.8 when are building against UCX 1.7.

File: pkgs/development/libraries/openmpi/default.nix (hunk 4 above)
Comment (markuskowa) [reply pkgs/development/libraries/openmpi/default.nix]: I would leave the whole package out for now since the `libfabric`  is not even built with OmniPath in the present version. We can add that later, when it becomes relevant.

File: pkgs/development/libraries/openmpi/default.nix (hunk 10 above)
Comment (ikervagyok) [reply pkgs/development/libraries/openmpi/default.nix]: Take a look at the two commit messages:

https://github.com/open-mpi/ompi/commit/526775dfd7ad75c308532784de4fb3ffed25458f
https://github.com/open-mpi/ompi/commit/a3026c016a6a8be379f62585b6ddc070175c8106

The first one adds support and the second replaces the version bound `version > 1.7` with `version >= 1.7`. They are named that way, to make it clear, that they can be removed together.

Comment (markuskowa) [reply pkgs/development/libraries/openmpi/default.nix]: The way I understand the patches the first one fixes a problem with ucx/master (correct me if I am wrong). Since we are not building against ucx 1.8 (or current master) that patch is not needed right now. We should remove it for now.

**Comment by markuskowa** [comment]
Can we close this PR? Most of the improvements have now been addressed in other PRs.
